### PR TITLE
Remove some contexts and simplify storage code

### LINKED ISSUE #134: 3.7-build Actions hang on Run Tests
On #132, all checks aside from 3.7-build passed within ~10 minutes. The 3.7-build checks hung on the Run Tests step until they eventually timed out. I just restarted them but they appear to be hanging again.

### LINKED ISSUE #144: Intermittent hanging when opening cached files without failure.
In recent attempts at running the [SMAP SSS Recipe](https://github.com/pangeo-forge/staged-recipes/pull/31) we have noted intermittent hanging when opening from cache via [_maybe_open_or_copy_to_local](https://github.com/pangeo-forge/pangeo-forge-recipes/blob/0.3.3/pangeo_forge_recipes/recipes/xarray_zarr.py#L61).  These are dask worker log entries from an recently attempted run where workers were hung

```
INFO:pangeo_forge_recipes.recipes.xarray_zarr:Opening 'https://podaac-opendap.jpl.nasa.gov/opendap/allData/smap/L3/RSS/V4/monthly/SCI/2015//RSS_smap_SSS_L3_monthly_2015_08_FNL_v04.0.nc' from cache
```
```
INFO:pangeo_forge_recipes.recipes.xarray_zarr:Opening 'https://podaac-opendap.jpl.nasa.gov/opendap/allData/smap/L3/RSS/V4/monthly/SCI/2015//RSS_smap_SSS_L3_monthly_2015_11_FNL_v04.0.nc' from cache
```
Interestingly, though the files which hang for cache retrieval are different between tests but consistently 3 workers seem to hang.  Not sure if this indicates some type of concurrency issue with cache retrieval.

Besides blocking the recipe, this is problematic due to the asynchronous nature of these operations which can potentially leave workers in an idle healthy state and prevent cleanup from the scheduler.

@rabernat I had look over the storage refactoring [PR](https://github.com/pangeo-forge/pangeo-forge-recipes/commit/32e920185934e42baf32650c5b3d1e76ea078ab3#diff-e12c886cc124886c5cfa5313d760a36c39649af9da845077c663e6feab8487b5L388) but I'm unsure if this might potentially address this issue?  Do you have recommendations on any additional logging or telemetry you would like me to collect that might be helpful in diagnosing the root cause for this?

If you feel the storage refactoring PR might address this issue, can we craft a new release which includes this so it can be deployed in our `staged-recipes` CI workflows.

### PR DESCRIPTION
**Pull request by martindurant**
Fixes #144 
Fixes #134

### CODE REVIEW

**Comment by martindurant** [comment]
I explicitly tell the various instances of `xr.open_dataset` to use driver "h5netcdf", since sometimes it guesses netcdf4, which fails for some file-like sometimes.

**Comment by rabernat** [comment]
Interesting and useful suggestion. Curious to see if it works.

If it works, we should think carefully about the intersection with `xarray_open_kwargs`. For some recipes, the user will specify an engine (e.g. `engine='cfgrib'`) in order to read special formats.

File: ci/py3.8.yml [hunk 1]
```diff
@@ -10,7 +10,7 @@ dependencies:
   - codecov
   - dask
   - distributed
-  # - fsspec>=2021.04.0
+  - fsspec>=2021.04.0
```
Comment (rabernat) [inline ci/py3.8.yml:13]: Didn't I already do this in #148?

Comment (martindurant) [reply ci/py3.8.yml:13]: I must not have pulled/merged.

**Comment by martindurant** [comment]
> For some recipes, the user will specify an engine (e.g. engine='cfgrib')

Those would presumably *require* copy-to-local.

**Comment by rabernat** [comment]
Did this really just fix it?

**Comment by rabernat** [comment]
> Those would presumably _require_ copy-to-local.

YES

**Comment by rabernat** [comment]
I'll restart the tests just to get another datapoint.

**Comment by martindurant** [comment]
> Did this really just fix it?

I'll wait for more data points.

**Comment by martindurant** [comment]
Is that a green?

**Comment by rabernat** [comment]
Btw, didn't I make an fsspec issue about this "double context manager" thing? I can't seem to find it anymore.

It's really an xarray bug IMO.

**Comment by martindurant** [comment]
The  LocalFileOpener became a derivative of io.IOBase in 65880ef8dbeadfd655a5aefb5436f2390042d837 , so it shouldn't be necessary any more; but for local files that netcdf4 can normally cope with, it still fails. I guess there is an exact instance check somewhere, but I can't see where. Perhaps it's tied up with how xarray guesses file types by looking at magic bytes. I haven't checked whether netcdf4 itself can cope with LocalFileOpener or if it's only via xarray.

**Comment by martindurant** [comment]
https://github.com/intake/filesystem_spec/pull/589

**Comment by rabernat** [comment]
I'm gonna merge this, ok?

**Comment by martindurant** [comment]
Yep

**Pull request merged**